### PR TITLE
[IREE] Inference & Benchmark

### PR DESCRIPTION
**Pull request by Vladislav-Denisov**

### CODE REVIEW

File: src/inference/inference_iree.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    import iree.runtime as ireert  # noqa: E402
+except ImportError as e:
+    log.error(f"IREE import error: {e}")
+    sys.exit(1)
+
+
+def cli_argument_parser():
+    parser = argparse.ArgumentParser()
+
+
```
Comment (valentina-kustikova) [inline src/inference/inference_iree.py]: Лишняя пустая строка.

File: src/inference/README.md [hunk 2]
```diff
@@ -1492,3 +1569,4 @@ python inference_ncnn.py --model <model_name> \
 [dgl]: https://www.dgl.ai/pages/start.html
 [ogb]: https://ogb.stanford.edu/
 [tensorflow-gpu]: https://www.tensorflow.org/install/pip
+[iree]: https://iree.dev/
```
Comment (valentina-kustikova) [inline src/inference/README.md]: Слэш на конце можно удалить.

File: src/inference/inference_iree.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+                        nargs=3,
+                        dest='std')
+    parser.add_argument('--channel_swap',
+                        help='Parameter of channel swap.',
+                        default=[2, 1, 0],
+                        type=int,
+                        nargs=3,
+                        dest='channel_swap')
+    parser.add_argument('-tb', '--target_backend',
+                        help='Target backend, for example "llvm-cpu" for CPU.',
```
Comment (valentina-kustikova) [inline src/inference/inference_iree.py]: Здесь точно не требуется экранирование двойных кавычек в строке?

File: src/inference/inference_iree.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+                        default='llvm-cpu',
+                        type=str,
+                        dest='target_backend')
+    parser.add_argument('--opt_level',
+                        help='The optimization level of the task extractions.',
```
Comment (valentina-kustikova) [inline src/inference/inference_iree.py]: Не очень поняла, что здесь имелось ввиду в комментарии `the task extractions`?

File: src/inference/inference_iree.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        log.info(f'Starting model compilation')
+        return IREECompiler.compile(mlir_path, target_backend, opt_level, extra_compile_args)
+    except Exception as e:
+        log.error(f"Failed to compile MLIR: {e}")
+        raise
+
+
+def load_iree_model(vmfb_buffer):
+    try:
+        config = ireert.Config('local-task')
```
Comment (valentina-kustikova) [inline src/inference/inference_iree.py]: 'local-task' - единственное возможное значение параметра функции? Не понятно, имеет ли смысл выносить в константы или параметры.

File: src/inference/inference_iree.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+
+    return parser.parse_args()
+
+
+def compile_mlir(mlir_path, target_backend, opt_level, extra_compile_args):
+    try:
+        log.info(f'Starting model compilation')
+        return IREECompiler.compile(mlir_path, target_backend, opt_level, extra_compile_args)
+    except Exception as e:
+        log.error(f"Failed to compile MLIR: {e}")
```
Comment (valentina-kustikova) [inline src/inference/inference_iree.py]: f-строка где-то в двойных кавычках, а где-то - в одинарных. Надо сделать единоообразно, как в других скриптах вывода.

File: src/inference/inference_iree.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+def load_iree_model(vmfb_buffer):
+    try:
+        config = ireert.Config('local-task')
+
+        vm_module = ireert.VmModule.from_flatbuffer(config.vm_instance, vmfb_buffer)
+        context = ireert.SystemContext(config=config)
+        context.add_vm_module(vm_module)
+
+        log.info(f"Successfully loaded IREE model")
```
Comment (valentina-kustikova) [inline src/inference/inference_iree.py]: Аналогичный комментарий касательно f-строки.

File: src/inference/inference_iree.py [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+        return context
+
+    except Exception as e:
+        log.error(f"Failed to load IREE model: {e}")
```
Comment (valentina-kustikova) [inline src/inference/inference_iree.py]: Аналогичный комментарий касательно f-строки. Посмотрите, пожалуйста, по всему коду.

File: src/inference/inference_iree.py [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+
+def inference_iteration(inference_func, get_slice):
+    slice_input = get_slice()
+    _, exec_time = infer_slice(inference_func, slice_input)
+    return exec_time
+
+
+@get_exec_time()
+def infer_slice(inference_func, slice_input):
+    config = ireert.Config('local-task')
```
Comment (valentina-kustikova) [inline src/inference/inference_iree.py:206]: Такой же вопрос по 'local-task'.

File: src/inference/inference_iree.py [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+            result = result.to_host()
+        
+        # Extract tensor from dict if needed
+        if isinstance(result, dict):
+            result_key = next(iter(result))
+            logits = result[result_key]
+            output_key = result_key
+        else:
+            logits = np.array(result)
+            output_key = 'output'
```
Comment (valentina-kustikova) [inline src/inference/inference_iree.py]: Поскольку константа используется только в return, то имеет смысл ее напрямую там и написать.

File: src/inference/inference_iree.py [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+        )
+
+        file_type = args.model.split('.')[-1]
+        if file_type == 'mlir':
+            vmfb_buffer = compile_mlir(args.model, args.target_backend, args.opt_level, args.extra_compile_args)
+        elif file_type == 'vmfb':
+            with open(args.model, 'rb') as f:
+                vmfb_buffer = f.read()
+        else:
+            raise ValueError(f'The file type {file_type} is not supported')
```
Comment (valentina-kustikova) [inline src/inference/inference_iree.py]: Предлагают вынести компиляцию/чтение модели в отдельную функцию.

File: src/inference/inference_iree.py [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+
+        report_writer = ReportWriter()
+        report_writer.update_framework_info(name='IREE')
+        report_writer.update_configuration_setup(
+            batch_size=args.batch_size,
+            iterations_num=args.number_iter,
+            target_device=args.target_backend
+        )
+
+        file_type = args.model.split('.')[-1]
```
Comment (valentina-kustikova) [inline src/inference/inference_iree.py]: Перед этим этапом надо добавить логирование, поскольку далее логируются все шаги основной функции.

File: src/inference/io_model_wrapper.py [hunk 13]
```diff
@@ -409,3 +409,19 @@ def get_input_layer_dtype(self):
...
     pass
+
+
+class IREEModelWrapper(IOModelWrapper):
+    def __init__(self, args):
+        self._input_shapes = [args.input_shape]
+        self._model_path = args.model
+    
+    def get_input_layer_names(self, model):
+        return ['input']
```
Comment (valentina-kustikova) [inline src/inference/io_model_wrapper.py:420]: А здесь точно для всех моделей одинаковое название входного слоя будет? Надо подумать, но скорее всего название входного слоя придется отдавать через параметры.

File: src/model_converters/iree_converter/README.md [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+- Required: `--source_framework pytorch`, `--model_name <model_name>`, `--torch_module <module>`, `--input_shape B W H C`, `--output_mlir <output_path>`
+- Optional: `--weights <path/to/weights.pth>`
+
+### Examples of usage
+ONNX model conversion:
+```sh
+python3 iree_converter.py -f onnx -m efficientnet-b0.onnx \
+                         --onnx_opset_version 18 \
+                         -o ./output/efficientnet-b0.mlir
+```
```
Comment (valentina-kustikova) [inline src/model_converters/iree_converter/README.md:54]: Здесь имеет смысл добавить в примечании ссылку, откуда была загружена модель.

File: src/model_converters/iree_converter/README.md [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+PyTorch model from file:
+```sh
+python3 iree_converter.py -f pytorch -m resnet50.pt \
+                         -is 1 224 224 3 \
+                         -o ./output/resnet50.mlir
+```
```
Comment (valentina-kustikova) [inline src/model_converters/iree_converter/README.md:61]: Здесь имеет смысл добавить в примечании ссылку, откуда была загружена модель.

File: src/model_converters/iree_converter/README.md [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+iree_compiler.py --mlir <input.mlir> \
+                 --target_backend <target_backend> \
+                 --opt_level <opt_level> \
+                 --output_file <output_file> \
+                 [--extra_args <extra_args>]
+```
+This script compiles model from `.mlir` format to the deployable binary format for the specified target backend.
+
+### IREE compiler parameters
+- `-m / --mlir` - Path to an .mlir file with a model. Required.
```
Comment (valentina-kustikova) [inline src/model_converters/iree_converter/README.md]: Во всем списке без тире пишем `...is a path...`

File: src/model_converters/iree_converter/README.md [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+- `-tb / --target_backend` - Target backend for compilation. Required. Examples: `llvm-cpu`, `cuda`, `vulkan`, `vmvx`.
+- `--opt_level` - The optimization level of the compilation. Choices: `0`, `1`, `2`, `3`. Default: `2`.
+- `-o / --output_file` - Path to save the compiled model. Required.
+- `--extra_args` - Extra arguments for compilation. Optional.
+
+### Supported target backends
+- `llvm-cpu` - CPU execution using LLVM
```
Comment (valentina-kustikova) [inline src/model_converters/iree_converter/README.md]: Во всем списке без тире пишем `...is a path...` и точка на конце предложения.

**Review by valentina-kustikova — COMMENTED** [review]
@Vladislav-Denisov, посмотрела код, написала небольшие комментарии. Надо еще отследить, что проверка стиля кодирования везде проходит.

File: src/inference/io_model_wrapper.py (hunk 13 above)
Comment (Vladislav-Denisov) [reply src/inference/io_model_wrapper.py:420]: Для IREE на самом деле не требуется наименование входного слоя.
Он использует наименование функции, в скомпилированном модуле, которую потребуется исполнить https://github.com/Vladislav-Denisov/dl-benchmark/blob/70526453d3c0421d93746556507e7c6096ed0cbf/src/inference/inference_iree.py#L59

Входной слой захардкожен, потому что он требуется для некоторых методов [IOAdapter](https://github.com/Vladislav-Denisov/dl-benchmark/blob/70526453d3c0421d93746556507e7c6096ed0cbf/src/inference/io_adapter.py#L12)

File: src/inference/inference_iree.py (hunk 10 above)
Comment (Vladislav-Denisov) [reply src/inference/inference_iree.py]: `output_key` используется также в https://github.com/Vladislav-Denisov/dl-benchmark/blob/70526453d3c0421d93746556507e7c6096ed0cbf/src/inference/iree_auxiliary.py#L197

Для универсального возврата в https://github.com/Vladislav-Denisov/dl-benchmark/blob/70526453d3c0421d93746556507e7c6096ed0cbf/src/inference/iree_auxiliary.py#L213

File: src/inference/inference_iree.py (hunk 5 above)
Comment (Vladislav-Denisov) [reply src/inference/inference_iree.py]: Все доступные варианты помимо `local-task` описаны [в табличке здесь](https://iree.dev/guides/deployment-configurations/#runtime-hal-drivers-and-devices)
Я захардкодил, потому что пока что рассматривал только запуск на CPU.

File: src/inference/README.md [hunk 18]
```diff
@@ -1482,6 +1483,124 @@ python inference_ncnn.py --model <model_name> \
...
+- `-fn / --function_name` - имя функции внутри IREE-модуля, которое будет вызвано при инференсе.
+- `-i / --input` - путь до изображения или директории с изображениями
+  (расширения файлов `.jpg`, `.png`, `.bmp` и т.д.).
+- `-is / --input_shape` - размеры входного тензора сети в формате
+  BxCxHxW, B - размер пачки, C - количество каналов изображений,
+  H - высота изображений, W - ширина изображений.
+
+#### Остальные параметры зависят от того, в каком формате модель подается на вход.
+
+
```
Comment (valentina-kustikova) [inline src/inference/README.md]: Лишняя пустая строка.

File: src/inference/README.md [hunk 19]
```diff
@@ -1482,6 +1483,124 @@ python inference_ncnn.py --model <model_name> \
...
+- **Готовый IREE-модуль (`.vmfb` или `.mlir`)**
+  - `-m / --model` - путь до модели в формате `.vmfb` (готовый бинарник) или `.mlir` (будет скомпилирован перед запуском), обязателен.
```
Comment (valentina-kustikova) [inline src/inference/README.md]: Вместо "обязателен" предлагаю везде написать "Обязательный параметр."

File: src/inference/iree_auxiliary.py [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+            logits = logits.reshape(1, -1)
+        elif logits.ndim > 2:
+            logits = logits.reshape(logits.shape[0], -1)
+
+        # Apply softmax
+        max_logits = np.max(logits, axis=-1, keepdims=True)
+        exp_logits = np.exp(logits - max_logits)
+        probabilities = exp_logits / np.sum(exp_logits, axis=-1, keepdims=True)
+
+        return {output_key: probabilities}
```
Comment (valentina-kustikova) [inline src/inference/iree_auxiliary.py]: Думаю, что это должно быть оформлен в отдельном классе, который является наследником `IOAdapter` или наследником адаптера для задачи классификации.

**Review by valentina-kustikova — COMMENTED** [review]
@Vladislav-Denisov, информацию о поддержке IREE надо добавить в основной README в корне репозитория. Также упали тесты с paddlepaddle, поправьте, пожалуйста, версию библиотеки.

File: src/inference/iree_auxiliary.py (hunk 20 above)
Comment (Vladislav-Denisov) [reply src/inference/iree_auxiliary.py]: Я реализовал `prepare_output` именно как отдельный вспомогательный метод, применяя подход аналогичный любому другому фреймворку , который требует подготовки выхода перед его передачей в
https://github.com/itlab-vision/dl-benchmark/blob/ec4106e285867e268bd1218dbd197eb9b1a7fc4c/src/inference/io_adapter.py#L246 

Например:
- PyTorch https://github.com/itlab-vision/dl-benchmark/blob/ec4106e285867e268bd1218dbd197eb9b1a7fc4c/src/inference/inference_pytorch.py#L545
- NCNN https://github.com/itlab-vision/dl-benchmark/blob/ec4106e285867e268bd1218dbd197eb9b1a7fc4c/src/inference/ncnn_auxiliary.py#L57
- MXNET https://github.com/itlab-vision/dl-benchmark/blob/ec4106e285867e268bd1218dbd197eb9b1a7fc4c/src/inference/mxnet_auxiliary.py#L16
- TensorFlow https://github.com/itlab-vision/dl-benchmark/blob/ec4106e285867e268bd1218dbd197eb9b1a7fc4c/src/inference/inference_tensorflow.py#L171
- и т.д.

Думаю, что если переделывать это на некоторый класс, который будет как-то связан с IOAdapter, то это стоит делать везде и это займет относительно много времени для аккуратного решения.

Пока что просто упростил `prepare_output` для IREE, избавился от лишнего.

File: src/inference/inference_iree.py (hunk 10 above)
Comment (Vladislav-Denisov) [reply src/inference/inference_iree.py]: Неактуально после https://github.com/itlab-vision/dl-benchmark/pull/590/commits/8044518875d1ea3f71749dfb25b210162c23bd62

File: README.md [hunk 21]
```diff
@@ -158,6 +160,9 @@ Please consider citing the following papers.
     - [`validation_results_tvm.md`](results/validation/validation_results_tvm.md)
       is a table that confirms correctness of inference implementation
       based on Apache TVM for several public models.
+    - [`validation_results_iree.md`](results/validation/validation_results_iree.md)
```
Comment (valentina-kustikova) [inline README.md:163]: Помимо этого файла надо добавить `results/iree_models_checklist.md` по аналогии с другими фреймворками.

File: README.md [hunk 22]
```diff
@@ -33,6 +33,7 @@ DLI supports inference using the following frameworks:
 - [ncnn][ncnn] (Python API).
 - [PaddlePaddle][PaddlePaddle] (Python API).
 - [ExecuTorch][executorch] (C++ and Python APIs)
+- [IREE][iree] (Python API)
```
Comment (valentina-kustikova) [inline README.md]: Добавьте, пожалуйста, точки в конце 35 и 36 строк.

**Comment by valentina-kustikova** [comment]
@Vladislav-Denisov, поправьте, пожалуйста, пару описанных выше моментов, и я думаю, что можно будет влить в основную ветку.

**Review by valentina-kustikova — APPROVED** [review]

**Comment by valentina-kustikova** [comment]
@Vladislav-Denisov, посмотрела, все ок. Заливаю в основную ветку.
